examples/throughput/subscriber.cpp

```diff
@@ -0,0 +1,327 @@
+                 subprefix << "Defaults:\n" <<
+                 subprefix << "./subscriber 0 0 \"Throughput example\"\n" << std::flush;
+    return EXIT_FAILURE;
+  }
+
+  if (argc > 1)
+  {
+    maxCycles = static_cast<unsigned long long>(atoi (argv[1])); /* The number of times to output statistics before terminating */
+  }
+  if (argc > 2)
+  {
+    pollingDelay = std::chrono::milliseconds(atoi(argv[2])); /* The number of ms to wait between reads (0 = waitset, -1 = listener) */
+  }
+  if (argc > 3)
+  {
+    partitionName = argv[3]; /* The name of the partition */
+  }
+  return EXIT_SUCCESS;
+}
+
+template<typename T>
+unsigned long long do_take(dds::sub::DataReader<T>& rd)
+{
+  auto samples = rd.take();
+  unsigned long long valid_samples = 0;
+  for (const auto & s:samples)
+  {
+    if (!s.info().valid())
+      continue;
+
+    auto ih = s.info().publication_handle();
+    auto ct = s.data().count();
+    auto it = mp.find(ih);
+    /*check whether the received sequence number matches that which we expect*/
+    if (it != mp.end() &&
+        it->second != ct)
+      outOfOrder++;
+
+    valid_samples++;
+    total_bytes += s.data().payload().size();
+    mp[ih] = ct+1;
+  }
+
+  total_samples += valid_samples;
+  return valid_samples;
+}
+
+template<typename T>
+void process_samples(dds::sub::DataReader<T> &reader, size_t max_c)
+{
+  unsigned long long prev_bytes = 0;
+  unsigned long long prev_samples = 0;
+  unsigned long cycles = 0;
+  auto startTime = std::chrono::steady_clock::now();
+  auto prev_time = startTime;
+
+  dds::core::cond::WaitSet waitset;
+
+  dds::core::cond::StatusCondition sc = dds::core::cond::StatusCondition(reader);
+  sc.enabled_statuses(dds::core::status::StatusMask::data_available());
+  waitset.attach_condition(sc);
+
+  dds::core::Duration waittime =
+    dds::core::Duration::from_millisecs(pollingDelay.count() > 0 ? pollingDelay.count() : 100);
+
+  while (!done && (max_c == 0 || cycles < max_c))
+  {
+    dds::core::cond::WaitSet::ConditionSeq conditions;
+    bool wait_again = true;
+    try {
+      if (pollingDelay.count() == 0) {
+        conditions = waitset.wait(waittime);
+      } else {
+        wait_again = false;
+        std::this_thread::sleep_for(pollingDelay.count() > 0 ? pollingDelay : std::chrono::milliseconds(100));
+      }
+    } catch (const dds::core::TimeoutError &) {
+      done = true;
+    } catch (const dds::core::Exception &e) {
+      std::cout << subprefix << "Waitset encountered the following error: \"" << e.what() << "\".\n" << std::flush;
+      break;
+    } catch (...) {
+      std::cout << subprefix << "Waitset encountered an unknown error.\n" << std::flush;
+      break;
+    }
+
+    if (reader.subscription_matched_status().current_count() == 0)
+      done = true;
+
+    for (const auto &c:conditions)
+    {
+      if (c == sc)
+      {
+        wait_again = false;
```

Comment:
Am I correct in stating that if you blocked on the WaitSet, and it triggers, then you leave the loop? In other words, you never block more than once for incoming data? Why is that?

Reply:
You are correct, the trigger value on the status conditions being checked is missing here.
I will check this.